### PR TITLE
Fix flex attention block mask adjustment counts

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186186

_adjust_num_blocks_and_indices truncates ordered block indices when a BlockMask is adjusted to a smaller number of key/value blocks. The old recount logic clamped the previous count and then counted truncated index values less than that count. That confused a block's column value with its position in the ordered valid prefix, so filler entries outside the original valid prefix could be counted as valid after truncation.

Fix the helper to derive the adjusted count from the original valid prefix and the new column bound. Surviving entries are stable-packed before truncation so the returned ordered representation remains consistent even when the valid prefix is not sorted by column value. This keeps the fix in the block-mask adjustment representation instead of special-casing full-kv masks or the reported shape.

Benchmark Results:
A microbenchmark comparing the old helper to the fixed helper on B=8,H=8,R=64,C=128,new_rows=32,new_cols=64 for 200 iterations showed the added stable packing cost outside the attention kernel hot path:
- CPU old: median 424.573 us, mean 428.086 us, min 400.842 us, max 541.184 us
- CPU new: median 1188.106 us, mean 1265.558 us, min 1031.921 us, max 2289.992 us
- CUDA old: median 67.992 us, mean 69.249 us, min 64.717 us, max 91.458 us
- CUDA new: median 135.408 us, mean 140.009 us, min 131.578 us, max 214.563 us

Test Plan:
- python test/inductor/test_flex_attention.py -k test_adjust_num_blocks_and_indices_recounts_truncated_blocks
- python test/inductor/test_flex_attention.py -k test_block_mask_transform_preserves_dq_write_order
- lintrunner -a (failed on unrelated PYREFLY unexpected-keyword in torch/nn/modules/loss.py for linear_bias)
- lintrunner -a --skip PYREFLY torch/nn/attention/flex_attention.py test/inductor/test_flex_attention.py

Fixes #139117
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the only failing CI job was the TorchTitan feature integration test failing in torchcomms process-group resolution, which is unrelated to this flex-attention mask adjustment change. I reran the failed workflow and will check back once CI finishes.

**Comment by jansel** [comment]
My agent says the current new CI failure is in the HuggingFace inductor job while importing DebertaV2ForMaskedLM in eager mode from Transformers, before this PR's flex-attention change is exercised. I reran the failed workflow and will check back once CI has a fresh result.

**Comment by jansel** [comment]
My agent says the repeated HuggingFace inductor failure was still the DebertaV2ForMaskedLM eager import regression outside this flex-attention change. I rebased the PR onto current main and resubmitted; the focused flex-attention tests and lintrunner -a pass locally, and CI is running again.

**Comment by jansel** [comment]
My agent says this is fixed by https://github.com/pytorch/pytorch/pull/188208, which landed as commit 5b86a2db4bd9 and closed #139117. Closing this duplicate PR.
